lib/instrumentation/modules/@aws-sdk/client-s3.js

```diff
@@ -0,0 +1,207 @@
+
+          if (input.Key) {
+            otelAttrs['aws.s3.key'] = input.Key
+          }
+        }
+
+        let err
+        let result
+        try {
+          result = await next(args)
+        } catch (ex) {
+          // Save the error for use in `finally` below, but re-throw it to
+          // not impact code flow.
+          err = ex
+          throw ex
+        } finally {
+          if (span) {
+            let statusCode
+            if (result && result.response) {
+              statusCode = result.response.statusCode
+            } else if (err && err.$metadata && err.$metadata.httpStatusCode) {
+              // This code path happens with a GetObject conditional request
+              // that returns a 304 Not Modified.
+              statusCode = err.$metadata.httpStatusCode
+            }
+            if (statusCode) {
+              span._setOutcomeFromHttpStatusCode(statusCode)
+            } else {
+              span._setOutcomeFromErrorCapture(constants.OUTCOME_FAILURE)
+            }
+            if (err && (!statusCode || statusCode >= 400)) {
+              agent.captureError(err, { skipOutcome: true })
+            }
+
+            // Configuring `new S3Client({useArnRegion:true})` allows one to
+            // use an Access Point bucket ARN for a region *other* than the
+            // one for which the client is configured. Therefore, we attempt
+            // to get the bucket region from the ARN first.
+            const useArnRegion = await client.config.useArnRegion()
+            const region = useArnRegion && args.input.Bucket.startsWith('arn:')
+              ? regionFromS3Arn(args.input.Bucket)
+              : await client.config.region()
+
+            // Destination context.
+            let port = context[elasticAPMStash].port
+            if (port === undefined) {
+              if (context[elasticAPMStash].protocol === 'https:') {
+                port = 443
+              } else if (context[elasticAPMStash].protocol === 'http:') {
+                port = 80
+              }
+            }
+            const destContext = {
+              address: context[elasticAPMStash].hostname,
+              port: port,
+              service: {
+                name: SUBTYPE,
+                type: TYPE
+              }
+            }
+            if (resource) {
+              destContext.service.resource = resource
+            }
+
+            if (region) {
+              destContext.cloud = { region }
+            }
+            span._setDestinationContext(destContext)
+
+            span.end()
+          }
+        }
+
+        return result
+      },
+      options: { step: 'initialize', priority: 'high', name: 'elasticAPMSpan' }
+    },
+    {
+      middleware: (next, context) => async (args) => {
+        context[elasticAPMStash] = {
+          hostname: args.request.hostname,
+          port: args.request.port,
+          protocol: args.request.protocol
+        }
+        return await next(args)
```

Comment:
I'm no longer confident that the right way to write a aws-sdk middleware is to `await` the `next(args)` call. Do you know of a reference for that?  Looking at some examples for "middleware-sdk-s3":

- this one does *not* await: https://github.com/aws/aws-sdk-js-v3/blob/main/packages/middleware-sdk-s3/src/check-content-length-header.ts#L40
- this one *does* await: https://github.com/aws/aws-sdk-js-v3/blob/main/packages/middleware-sdk-s3/src/throw-200-exceptions.ts#L18

It would be good to have some comfort around that. For example, does this instrumentation work fine if this `await` is dropped?   I assume the await above in the "initialize" middleware is necessary.

Reply:
I think this is a leftover. It doesn't fail but I see no point on `await`ing a value to just return it. Next middleware will `then` on the returned promise.

Indeed the `await` on the prev middleware is necessary because we need the resolved result to extract info for the span.